src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java

```diff
@@ -26,48 +28,106 @@
 public class BuildNameSetter extends BuildWrapper implements MatrixAggregatable {
 
     public final String template;
-
+    public final String matrixTemplate;
+    private boolean matches;
+    private int scenario = -1;
+    private boolean started = false;
+    
     @DataBoundConstructor
+    public BuildNameSetter(String template, String matrixTemplate) {
+        this.template = template;
+        this.matrixTemplate = matrixTemplate;
+        matches = template.equals(matrixTemplate);
+        
+        if(matches || matrixTemplate.length()>0 && template.length()==0){
+        	scenario = 0;
+        }else if (matrixTemplate.length()>0 && template.length()>0){
+        	scenario = 1;
+        }else if (matrixTemplate.length()==0 && template.length()>0){
+        	scenario = 2;
+        }
+    }
+    
+    @Deprecated
     public BuildNameSetter(String template) {
         this.template = template;
+        this.matrixTemplate = template;
     }
-
+    
     @Override
     public Environment setUp(AbstractBuild build, Launcher launcher, BuildListener listener) throws IOException, InterruptedException {
-        setDisplayName(build, listener);
-
-        return new Environment() {
+        try{
+        	//not checking before setting name because this check is done via Jelly call to isMatrix()
+        	setDisplayName(build, listener);
+        }catch(Exception e){
+        	e.printStackTrace();
+        	return null;
+        }
+    	return new Environment() {
             @Override
             public boolean tearDown(AbstractBuild build, BuildListener listener) throws IOException, InterruptedException {
-                setDisplayName(build, listener);
-                return true;
+                try {
+					setDisplayName(build, listener);
+	                return true;
+				} catch (Exception e) {
+					e.printStackTrace();
+				}
+                return false;
             }
-        };
+    	};
     }
 
-    private void setDisplayName(AbstractBuild build, BuildListener listener) throws IOException, InterruptedException {
-        try {
-            build.setDisplayName(TokenMacro.expandAll(build, listener, template));
-        } catch (MacroEvaluationException e) {
-            listener.getLogger().println(e.getMessage());
-        }
+    private void setDisplayName(AbstractBuild build, BuildListener listener) throws Exception {
+    	try{
+    		switch(scenario){
+    		case 0:
+    	    		build.setDisplayName(TokenMacro.expandAll(build, listener, matrixTemplate));
+    				return;
+    		case 1:
+	    			if(started){	
+	    				build.setDisplayName(TokenMacro.expandAll(build, listener, matrixTemplate));
+	    			}else{
+	    				build.setDisplayName(TokenMacro.expandAll(build, listener, template));
+	    			}
+    				return;
+    		case 2:
+	    			build.setDisplayName(TokenMacro.expandAll(build, listener, template));
+    				return;
+    		case -1:
+    				throw new Exception("No valid values set for build names!");
```

Comment:
Should probably not be throwing a new `Exception`, but instead `java.lang.IllegalArgumentException`. Then the `private void setDisplayName(AbstractBuild build, BuildListener listener) throws Exception` should be changed to throw `IllegalArgumentException`.